src/fcntl.rs

```diff
@@ -53,6 +55,23 @@ pub fn openat<P: ?Sized + NixPath>(dirfd: RawFd, path: &P, oflag: OFlag, mode: M
     Errno::result(fd)
 }
 
+pub fn readlink<'a, P: ?Sized + NixPath>(path: &P, buffer: &'a mut [u8]) -> Result<&'a OsStr> {
+    let res = try!(path.with_nix_path(|cstr| {
+        unsafe { libc::readlink(cstr.as_ptr(), buffer.as_mut_ptr() as *mut c_char, buffer.len() as size_t) }
+    }));
+
+    Errno::result(res).map(move |len| OsStr::from_bytes(&buffer[..(len as usize)]))
```

Comment:
I decided to use strategy 2 for the moment. If people want to use the first option, they could also use the rust std.